File: apps/server/providers/src/usage/claude.ts
import type {
  UsageAmount,
  UsageFetchContext,
  UsageFetchParams,
  UsageLimit,
  UsageProvider,
  UsageReport,
  UsageWindow,
} from "@console/types";
import { CLAUDE_CODE_VERSION } from "@/providers/src/claude/constants.js";
import { HOUR_MS, WEEK_MS, parseIsoTimestamp } from "./shared.js";

const DEFAULT_ENDPOINT = "https://api.anthropic.com/api/oauth";

const CLAUDE_HEADERS = {
  accept: "application/json, text/plain, */*",
  "content-type": "application/json",
  "user-agent": `claude-cli/${CLAUDE_CODE_VERSION} (external, cli)`,
  "anthropic-beta":
    "claude-code-20250219,oauth-2025-04-20,interleaved-thinking-2025-05-14,context-management-2025-06-27",
} as const;

interface ClaudeUsageBucket {
  utilization?: unknown;
  resets_at?: unknown;
}

interface ParsedBucket {
  utilization?: number;
  resetsAt?: number;
}

interface ClaudeApiLimitEntry {
  kind?: unknown;
  percent?: unknown;
  resets_at?: unknown;
  scope?: unknown;
}

interface ParsedLimitEntry {
  kind: string;
  bucket: ParsedBucket;
  displayName?: string;
}

function toNumber(value: unknown): number | undefined {
  if (typeof value === "number" && Number.isFinite(value)) return value;
  if (typeof value === "string" && value.trim() !== "") {
    const n = Number(value);
    if (Number.isFinite(n)) return n;
  }
  return undefined;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function parseBucket(bucket: unknown): ParsedBucket | undefined {
  if (!isRecord(bucket)) return undefined;
  const utilization = toNumber(bucket.utilization);
  const resetsAt = parseIsoTimestamp(typeof bucket.resets_at === "string" ? bucket.resets_at : undefined);
  if (utilization === undefined && resetsAt === undefined) return undefined;
  return { utilization, resetsAt };
}

function displayNameOf(scope: unknown): string | undefined {
  if (!isRecord(scope)) return undefined;
  const model = scope.model;
  if (!isRecord(model)) return undefined;
  const name = model.display_name;
  return typeof name === "string" && name.trim() ? name.trim() : undefined;
}

function parseLimitEntries(raw: unknown): ParsedLimitEntry[] {
  if (!Array.isArray(raw)) return [];
  const entries: ParsedLimitEntry[] = [];
  for (const item of raw) {
    if (!isRecord(item)) continue;
    const entry = item as ClaudeApiLimitEntry;
    if (typeof entry.kind !== "string") continue;
    const utilization = toNumber(entry.percent);
    const resetsAt = parseIsoTimestamp(typeof entry.resets_at === "string" ? entry.resets_at : undefined);
    if (utilization === undefined && resetsAt === undefined) continue;
    const displayName = displayNameOf(entry.scope);
    entries.push({
      kind: entry.kind,
      bucket: { utilization, resetsAt },
      ...(displayName ? { displayName } : {}),
    });
  }
  return entries;
}

function normalizeBaseUrl(baseUrl?: string): string {
  const trimmed = baseUrl?.trim().replace(/\/+$/, "");
  if (!trimmed) return DEFAULT_ENDPOINT;
  if (trimmed.toLowerCase().endsWith("/api/oauth")) return trimmed;
  try {
    const url = new URL(trimmed);
    let path = url.pathname.replace(/\/+$/, "");
    if (path === "/") path = "";
    if (path.toLowerCase().endsWith("/v1")) path = path.slice(0, -3);
    if (!path) return `${url.origin}/api/oauth`;
    return `${url.origin}${path}/api/oauth`;
  } catch {
    return DEFAULT_ENDPOINT;
  }
}

function buildAmount(utilization: number | undefined): UsageAmount {
  if (utilization === undefined) return { unit: "percent" };
  const clamped = Math.min(Math.max(utilization, 0), 100);
  const usedFraction = clamped / 100;
  return {
    used: clamped,
    limit: 100,
    remaining: Math.max(0, 100 - clamped),
    usedFraction,
    remainingFraction: Math.max(0, 1 - usedFraction),
    unit: "percent",
  };
}

function buildStatus(usedFraction: number | undefined): UsageLimit["status"] {
  if (usedFraction === undefined) return "unknown";
  if (usedFraction >= 1) return "exhausted";
  if (usedFraction >= 0.5) return "warning";
  return "ok";
}

function buildLimit(args: {
  id: string;
  label: string;
  windowId: string;
  windowLabel: string;
  durationMs: number;
  bucket: ParsedBucket | undefined;
  provider: UsageLimit["scope"]["provider"];
  tier?: string;
  shared?: boolean;
}): UsageLimit | null {
  if (!args.bucket) return null;
  const amount = buildAmount(args.bucket.utilization);
  if (amount.used === undefined) return null;
  const window: UsageWindow = {
    id: args.windowId,
    label: args.windowLabel,
    durationMs: args.durationMs,
    ...(args.bucket.resetsAt !== undefined ? { resetsAt: args.bucket.resetsAt } : {}),
  };
  return {
    id: args.id,
    label: args.label,
    scope: {
      provider: args.provider,
      windowId: args.windowId,
      ...(args.tier !== undefined ? { tier: args.tier } : {}),
      ...(args.shared !== undefined ? { shared: args.shared } : {}),
    },
    window,
    amount,
    status: buildStatus(amount.usedFraction),
  };
}

function slugify(displayName: string): string {
  return displayName
    .trim()
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-+|-+$/g, "");
}

function parseDollarAmount(amountMinor: unknown, exponent: unknown, currency: unknown): number | undefined {
  if (
    typeof amountMinor !== "number" ||
    !Number.isSafeInteger(amountMinor) ||
    amountMinor < 0 ||
    typeof exponent !== "number" ||
    !Number.isSafeInteger(exponent) ||
    exponent < 0
  ) {
    return undefined;
  }
  if (currency !== undefined && (typeof currency !== "string" || currency.toUpperCase() !== "USD")) {
    return undefined;
  }
  const divisor = 10 ** exponent;
  if (!Number.isFinite(divisor)) return undefined;
  const dollars = amountMinor / divisor;
  return Number.isFinite(dollars) ? dollars : undefined;
}

function parseSpendExtra(value: unknown): { used: number; limit?: number } | null {
  if (!isRecord(value) || value.enabled !== true || !Object.hasOwn(value, "limit") || !isRecord(value.used)) {
    return null;
  }
  const used = parseDollarAmount(value.used.amount_minor, value.used.exponent, value.used.currency);
  if (used === undefined) return null;
  if (value.limit === null) return { used };
  if (!isRecord(value.limit)) return null;
  const limit = parseDollarAmount(value.limit.amount_minor, value.limit.exponent, value.limit.currency);
  return limit === undefined || limit <= 0 ? null : { used, limit };
}

function parseLegacyExtra(value: unknown): { used: number; limit?: number } | null {
  if (!isRecord(value) || value.is_enabled !== true || !Object.hasOwn(value, "monthly_limit")) return null;
  const decimalPlaces = value.decimal_places === undefined ? 2 : value.decimal_places;
  const used = parseDollarAmount(value.used_credits, decimalPlaces, value.currency);
  if (used === undefined) return null;
  if (value.monthly_limit === null || value.monthly_limit === undefined) return { used };
  const limit = parseDollarAmount(value.monthly_limit, decimalPlaces, value.currency);
  return limit === undefined || limit <= 0 ? null : { used, limit };
}

function buildExtraLimit(
  payload: Record<string, unknown>,
  provider: UsageLimit["scope"]["provider"],
): UsageLimit | null {
  const raw = payload.spend === null || payload.spend === undefined ? payload.extra_usage : payload.spend;
  const parsed =
    payload.spend === null || payload.spend === undefined ? parseLegacyExtra(raw) : parseSpendExtra(raw);
  if (!parsed) return null;
  if (parsed.limit === undefined) {
    return {
      id: "claude:extra",
      label: "Claude Extra Usage",
      scope: { provider, windowId: "extra" },
      amount: { used: parsed.used, unit: "usd" },
    };
  }
  const remaining = Math.max(0, parsed.limit - parsed.used);
  const usedFraction = parsed.used / parsed.limit;
  const remainingFraction = remaining / parsed.limit;
  if (!Number.isFinite(usedFraction) || !Number.isFinite(remainingFraction)) return null;
  return {
    id: "claude:extra",
    label: "Claude Extra Usage",
    scope: { provider, windowId: "extra" },
    amount: {
      used: parsed.used,
      limit: parsed.limit,
      remaining,
      usedFraction,
      remainingFraction,
      unit: "usd",
    },
    status: parsed.used >= parsed.limit ? "exhausted" : buildStatus(usedFraction),
  };
}

function parseUnifiedWindow(
  headers: Record<string, string>,
  window: "5h" | "7d" | "7d_oi",
): ParsedBucket | undefined {
  const prefix = `anthropic-ratelimit-unified-${window}-`;
  const fraction = toNumber(headers[`${prefix}utilization`]);
  const resetSeconds = toNumber(headers[`${prefix}reset`]);
  const utilization = fraction === undefined ? undefined : fraction * 100;
  const resetsAt = resetSeconds !== undefined && resetSeconds > 0 ? resetSeconds * 1000 : undefined;
  if (utilization === undefined && resetsAt === undefined) return undefined;
  return { utilization, resetsAt };
}

export function parseClaudeRateLimitHeaders(
  headers: Record<string, string>,
  now = Date.now(),
): UsageReport | null {
  const fiveHour = parseUnifiedWindow(headers, "5h");
  const sevenDay = parseUnifiedWindow(headers, "7d");
  const scoped = parseUnifiedWindow(headers, "7d_oi");
  const limits = [
    buildLimit({
      id: "claude:5h",
      label: "Claude 5 Hour",
      windowId: "5h",
      windowLabel: "5 Hour",
      durationMs: 5 * HOUR_MS,
      bucket: fiveHour,
      provider: "claude",
      shared: true,
    }),
    buildLimit({
      id: "claude:7d",
      label: "Claude 7 Day",
      windowId: "7d",
      windowLabel: "7 Day",
      durationMs: WEEK_MS,
      bucket: sevenDay,
      provider: "claude",
      shared: true,
    }),
    buildLimit({
      id: "claude:7d:fable",
      label: "Claude 7 Day (Fable)",
      windowId: "7d",
      windowLabel: "7 Day",
      durationMs: WEEK_MS,
      bucket: scoped,
      provider: "claude",
      tier: "fable",
    }),
  ].filter((limit): limit is UsageLimit => limit !== null);
  if (limits.length === 0) return null;
  return { provider: "claude", fetchedAt: now, limits, metadata: { source: "ratelimit-headers" } };
}

export const claudeUsageProvider: UsageProvider = {
  id: "claude",
  supports: params => params.provider === "claude" && params.credential.type === "oauth",
  parseRateLimitHeaders: parseClaudeRateLimitHeaders,
  async fetchUsage(params: UsageFetchParams, ctx: UsageFetchContext): Promise<UsageReport | null> {
    if (params.provider !== "claude") return null;
    const { credential } = params;
    if (credential.type !== "oauth" || !credential.accessToken) return null;

    const nowMs = Date.now();
    if (credential.expiresAt !== undefined && credential.expiresAt <= nowMs) {
      ctx.logger?.warn("Claude usage token expired", { provider: params.provider });
      return null;
    }

    const baseUrl = normalizeBaseUrl(params.baseUrl);
    const url = `${baseUrl}/usage`;
    let payload: unknown;
    try {
      const response = await ctx.fetch(url, {
        headers: { ...CLAUDE_HEADERS, authorization: `Bearer ${credential.accessToken}` },
        signal: params.signal,
      });
      if (!response.ok) {
        ctx.logger?.warn("Claude usage request failed", { status: response.status, provider: params.provider });
        return null;
      }
      payload = await response.json();
    } catch (error) {
      ctx.logger?.warn("Claude usage request error", { provider: params.provider, error: String(error) });
      return null;
    }

    if (!isRecord(payload)) return null;
    const entries = parseLimitEntries(payload.limits);
    const fiveHour = parseBucket(payload.five_hour) ?? entries.find(entry => entry.kind === "session")?.bucket;
    const sevenDay = parseBucket(payload.seven_day) ?? entries.find(entry => entry.kind === "weekly_all")?.bucket;
    const sevenDayOpus = parseBucket(payload.seven_day_opus);
    const sevenDaySonnet = parseBucket(payload.seven_day_sonnet);

    const seenSlugs = new Set<string>();
    const scoped: UsageLimit[] = [];
    for (const entry of entries) {
      if (entry.kind !== "weekly_scoped" || !entry.displayName) continue;
      const slug = slugify(entry.displayName);
      if (!slug || seenSlugs.has(slug)) continue;
      seenSlugs.add(slug);
      const limit = buildLimit({
        id: `claude:7d:${slug}`,
        label: `Claude 7 Day (${entry.displayName})`,
        windowId: "7d",
        windowLabel: "7 Day",
        durationMs: WEEK_MS,
        bucket: entry.bucket,
        provider: params.provider,
        tier: slug,
      });
      if (limit) scoped.push(limit);
    }

    const limits = [
      buildLimit({
        id: "claude:5h",
        label: "Claude 5 Hour",
        windowId: "5h",
        windowLabel: "5 Hour",
        durationMs: 5 * HOUR_MS,
        bucket: fiveHour,
        provider: params.provider,
        shared: true,
      }),
      buildLimit({
        id: "claude:7d",
        label: "Claude 7 Day",
        windowId: "7d",
        windowLabel: "7 Day",
        durationMs: WEEK_MS,
        bucket: sevenDay,
        provider: params.provider,
        shared: true,
      }),
      buildLimit({
        id: "claude:7d:opus",
        label: "Claude 7 Day (Opus)",
        windowId: "7d",
        windowLabel: "7 Day",
        durationMs: WEEK_MS,
        bucket: sevenDayOpus,
        provider: params.provider,
        tier: "opus",
      }),
      buildLimit({
        id: "claude:7d:sonnet",
        label: "Claude 7 Day (Sonnet)",
        windowId: "7d",
        windowLabel: "7 Day",
        durationMs: WEEK_MS,
        bucket: sevenDaySonnet,
        provider: params.provider,
        tier: "sonnet",
      }),
      ...scoped,
      buildExtraLimit(payload, params.provider),
    ].filter((limit): limit is UsageLimit => limit !== null);

    if (limits.length === 0) return null;
    return {
      provider: params.provider,
      fetchedAt: nowMs,
      limits,
      metadata: {
        endpoint: url,
        ...(credential.accountId ? { accountId: credential.accountId } : {}),
        ...(credential.email ? { email: credential.email } : {}),
      },
      raw: payload,
    };
  },
};

File: apps/server-go/internal/providers/usage/claude.go
// Claude quota fetcher. Port of providers/src/usage/claude.ts: GET
// {base}/usage with the subscription OAuth token, parsed into a
// UsageReport with 5h/7d/scoped/extra limits.
package usage

import (
	"context"
	"encoding/json"
	"io"
	"math"
	"net/http"
	"strings"
	"time"

	"github.com/Adelodunpeter25/console/apps/server-go/internal/providers/claude"
)

const claudeDefaultEndpoint = "https://api.anthropic.com/api/oauth"
const hourMs = int64(60 * 60 * 1000)
const weekMs = int64(7 * 24 * 60 * 60 * 1000)

func claudeHeaders(accessToken string) map[string]string {
	return map[string]string{
		"Accept":         "application/json, text/plain, */*",
		"Content-Type":   "application/json",
		"User-Agent":     "claude-cli/" + claude.CodeVersion + " (external, cli)",
		"anthropic-beta": "claude-code-20250219,oauth-2025-04-20,interleaved-thinking-2025-05-14,context-management-2025-06-27",
		"Authorization":  "Bearer " + accessToken,
	}
}

// normalizeClaudeUsageBaseURL mirrors normalizeBaseUrl in the TS provider.
func normalizeClaudeUsageBaseURL(baseURL string) string {
	trimmed := strings.TrimRight(strings.TrimSpace(baseURL), "/")
	if trimmed == "" {
		return claudeDefaultEndpoint
	}
	lower := strings.ToLower(trimmed)
	if strings.HasSuffix(lower, "/api/oauth") {
		return trimmed
	}
	origin, path, ok := splitOriginPath(trimmed)
	if !ok {
		return claudeDefaultEndpoint
	}
	path = strings.TrimRight(path, "/")
	if strings.HasSuffix(strings.ToLower(path), "/v1") {
		path = path[:len(path)-3]
	}
	if path == "" {
		return origin + "/api/oauth"
	}
	return origin + path + "/api/oauth"
}

func splitOriginPath(raw string) (origin, path string, ok bool) {
	idx := strings.Index(raw, "://")
	if idx < 0 {
		return "", "", false
	}
	rest := raw[idx+3:]
	slash := strings.Index(rest, "/")
	if slash < 0 {
		return raw, "", true
	}
	return raw[:idx+3+slash], rest[slash:], true
}

func parseIsoTimestampMs(v any) *int64 {
	s, ok := v.(string)
	if !ok || s == "" {
		return nil
	}
	for _, layout := range []string{time.RFC3339Nano, time.RFC3339} {
		if t, err := time.Parse(layout, s); err == nil {
			ms := t.UnixMilli()
			return &ms
		}
	}
	return nil
}

// claudeBucket is a normalized utilization/reset pair.
type claudeBucket struct {
	utilization *float64
	resetsAtMs  *int64
}

func parseClaudeBucket(payload any) *claudeBucket {
	m, ok := Record(payload)
	if !ok {
		return nil
	}
	b := &claudeBucket{}
	if v, ok := Number(m["utilization"]); ok {
		b.utilization = &v
	}
	b.resetsAtMs = parseIsoTimestampMs(m["resets_at"])
	if b.utilization == nil && b.resetsAtMs == nil {
		return nil
	}
	return b
}

type claudeLimitEntry struct {
	kind        string
	bucket      *claudeBucket
	displayName string
}

func parseClaudeLimitEntries(raw any) []claudeLimitEntry {
	list, ok := raw.([]any)
	if !ok {
		return nil
	}
	var out []claudeLimitEntry
	for _, item := range list {
		m, ok := Record(item)
		if !ok {
			continue
		}
		kind, ok := m["kind"].(string)
		if !ok {
			continue
		}
		utilization, hasUtil := Number(m["percent"])
		resetsAt := parseIsoTimestampMs(m["resets_at"])
		if !hasUtil && resetsAt == nil {
			continue
		}
		bucket := &claudeBucket{resetsAtMs: resetsAt}
		if hasUtil {
			bucket.utilization = &utilization
		}
		entry := claudeLimitEntry{kind: kind, bucket: bucket}
		if scope, ok := Record(m["scope"]); ok {
			if model, ok := Record(scope["model"]); ok {
				if name, ok := model["display_name"].(string); ok && strings.TrimSpace(name) != "" {
					entry.displayName = strings.TrimSpace(name)
				}
			}
		}
		out = append(out, entry)
	}
	return out
}

// claudeBuildLimit mirrors the TS buildLimit: fixed window duration/label
// (not derived from payload seconds, unlike Codex), percent amount.
func claudeBuildLimit(id, label, windowID, windowLabel string, durationMs int64, bucket *claudeBucket, provider, tier string, shared bool) *Limit {
	if bucket == nil {
		return nil
	}
	amount := BuildPercentAmount(bucket.utilization)
	if amount.Used == nil {
		return nil
	}
	window := Window{ID: windowID, Label: windowLabel, DurationMs: &durationMs, ResetsAt: bucket.resetsAtMs}
	return &Limit{
		ID: id, Label: label,
		Scope:  Scope{Provider: provider, WindowID: windowID, Tier: tier, Shared: shared},
		Window: &window, Amount: amount,
		Status: StatusForFraction(amount.UsedFraction, nil, nil),
	}
}

func claudeSlugify(name string) string {
	var b strings.Builder
	prevDash := true
	for _, r := range strings.ToLower(name) {
		if (r >= 'a' && r <= 'z') || (r >= '0' && r <= '9') {
			b.WriteRune(r)
			prevDash = false
		} else if !prevDash {
			b.WriteRune('-')
			prevDash = true
		}
	}
	return strings.Trim(b.String(), "-")
}

func parseDollarAmount(amountMinor, exponent any, currency any) *float64 {
	amt, ok := Number(amountMinor)
	if !ok || amt < 0 || math.Trunc(amt) != amt {
		return nil
	}
	exp, ok := Number(exponent)
	if !ok || exp < 0 || math.Trunc(exp) != exp {
		return nil
	}
	if currency != nil {
		s, ok := currency.(string)
		if !ok || strings.ToUpper(s) != "USD" {
			return nil
		}
	}
	divisor := math.Pow(10, exp)
	if math.IsInf(divisor, 0) || divisor == 0 {
		return nil
	}
	dollars := amt / divisor
	if !isFinite(dollars) {
		return nil
	}
	return &dollars
}

func isFinite(f float64) bool {
	return !math.IsNaN(f) && !math.IsInf(f, 0)
}

type dollarBudget struct {
	used  float64
	limit *float64
}

func parseSpendExtra(v any) *dollarBudget {
	m, ok := Record(v)
	if !ok {
		return nil
	}
	if enabled, ok := m["enabled"].(bool); !ok || !enabled {
		return nil
	}
	if _, hasLimit := m["limit"]; !hasLimit {
		return nil
	}
	usedRec, ok := Record(m["used"])
	if !ok {
		return nil
	}
	used := parseDollarAmount(usedRec["amount_minor"], usedRec["exponent"], usedRec["currency"])
	if used == nil {
		return nil
	}
	if m["limit"] == nil {
		return &dollarBudget{used: *used}
	}
	limitRec, ok := Record(m["limit"])
	if !ok {
		return nil
	}
	limit := parseDollarAmount(limitRec["amount_minor"], limitRec["exponent"], limitRec["currency"])
	if limit == nil || *limit <= 0 {
		return nil
	}
	return &dollarBudget{used: *used, limit: limit}
}

func parseLegacyExtra(v any) *dollarBudget {
	m, ok := Record(v)
	if !ok {
		return nil
	}
	if enabled, ok := m["is_enabled"].(bool); !ok || !enabled {
		return nil
	}
	if _, hasLimit := m["monthly_limit"]; !hasLimit {
		return nil
	}
	decimalPlaces := any(2.0)
	if dp, ok := m["decimal_places"]; ok {
		decimalPlaces = dp
	}
	used := parseDollarAmount(m["used_credits"], decimalPlaces, m["currency"])
	if used == nil {
		return nil
	}
	if m["monthly_limit"] == nil {
		return &dollarBudget{used: *used}
	}
	limit := parseDollarAmount(m["monthly_limit"], decimalPlaces, m["currency"])
	if limit == nil || *limit <= 0 {
		return nil
	}
	return &dollarBudget{used: *used, limit: limit}
}

func claudeBuildExtraLimit(payload map[string]any, provider string) *Limit {
	spend, hasSpend := payload["spend"]
	var parsed *dollarBudget
	if !hasSpend || spend == nil {
		parsed = parseLegacyExtra(payload["extra_usage"])
	} else {
		parsed = parseSpendExtra(spend)
	}
	if parsed == nil {
		return nil
	}
	if parsed.limit == nil {
		return &Limit{
			ID: "claude:extra", Label: "Claude Extra Usage",
			Scope:  Scope{Provider: provider, WindowID: "extra"},
			Amount: Amount{Used: &parsed.used, Unit: "usd"},
		}
	}
	remaining := math.Max(0, *parsed.limit-parsed.used)
	usedFraction := parsed.used / *parsed.limit
	remainingFraction := remaining / *parsed.limit
	if !isFinite(usedFraction) || !isFinite(remainingFraction) {
		return nil
	}
	status := "exhausted"
	if parsed.used < *parsed.limit {
		status = StatusForFraction(&usedFraction, nil, nil)
	}
	return &Limit{
		ID: "claude:extra", Label: "Claude Extra Usage",
		Scope: Scope{Provider: provider, WindowID: "extra"},
		Amount: Amount{
			Used: &parsed.used, Limit: parsed.limit, Remaining: &remaining,
			UsedFraction: &usedFraction, RemainingFraction: &remainingFraction, Unit: "usd",
		},
		Status: status,
	}
}

// ParseClaudePayload converts a /usage payload to a Report.
func ParseClaudePayload(payload any, provider, accountID, email, endpoint string, nowMs int64) *Report {
	m, ok := Record(payload)
	if !ok {
		return nil
	}
	entries := parseClaudeLimitEntries(m["limits"])

	fiveHour := parseClaudeBucket(m["five_hour"])
	if fiveHour == nil {
		for _, e := range entries {
			if e.kind == "session" {
				fiveHour = e.bucket
				break
			}
		}
	}
	sevenDay := parseClaudeBucket(m["seven_day"])
	if sevenDay == nil {
		for _, e := range entries {
			if e.kind == "weekly_all" {
				sevenDay = e.bucket
				break
			}
		}
	}
	sevenDayOpus := parseClaudeBucket(m["seven_day_opus"])
	sevenDaySonnet := parseClaudeBucket(m["seven_day_sonnet"])

	seenSlugs := map[string]bool{}
	var scoped []Limit
	for _, e := range entries {
		if e.kind != "weekly_scoped" || e.displayName == "" {
			continue
		}
		slug := claudeSlugify(e.displayName)
		if slug == "" || seenSlugs[slug] {
			continue
		}
		seenSlugs[slug] = true
		if limit := claudeBuildLimit("claude:7d:"+slug, "Claude 7 Day ("+e.displayName+")", "7d", "7 Day", weekMs, e.bucket, provider, slug, false); limit != nil {
			scoped = append(scoped, *limit)
		}
	}

	var limits []Limit
	if l := claudeBuildLimit("claude:5h", "Claude 5 Hour", "5h", "5 Hour", 5*hourMs, fiveHour, provider, "", true); l != nil {
		limits = append(limits, *l)
	}
	if l := claudeBuildLimit("claude:7d", "Claude 7 Day", "7d", "7 Day", weekMs, sevenDay, provider, "", true); l != nil {
		limits = append(limits, *l)
	}
	if l := claudeBuildLimit("claude:7d:opus", "Claude 7 Day (Opus)", "7d", "7 Day", weekMs, sevenDayOpus, provider, "opus", false); l != nil {
		limits = append(limits, *l)
	}
	if l := claudeBuildLimit("claude:7d:sonnet", "Claude 7 Day (Sonnet)", "7d", "7 Day", weekMs, sevenDaySonnet, provider, "sonnet", false); l != nil {
		limits = append(limits, *l)
	}
	limits = append(limits, scoped...)
	if l := claudeBuildExtraLimit(m, provider); l != nil {
		limits = append(limits, *l)
	}

	if len(limits) == 0 {
		return nil
	}
	metadata := map[string]any{"endpoint": endpoint}
	if accountID != "" {
		metadata["accountId"] = accountID
	}
	if email != "" {
		metadata["email"] = email
	}
	return &Report{Provider: provider, FetchedAt: nowMs, Limits: limits, Metadata: metadata, Raw: payload}
}

// FetchClaudeUsage GETs the Claude /usage endpoint and parses the report.
// Returns (nil, nil) when not logged in, the token is expired, or the
// upstream call fails — matching the TS provider's null contract.
func FetchClaudeUsage(ctx context.Context, client *http.Client, baseURL, accessToken, email string, expiresAtMs int64) (*Report, error) {
	if accessToken == "" {
		return nil, nil
	}
	nowMs := time.Now().UnixMilli()
	if expiresAtMs > 0 && expiresAtMs <= nowMs {
		return nil, nil
	}
	if client == nil {
		client = &http.Client{Timeout: 5 * time.Second}
	}
	endpoint := normalizeClaudeUsageBaseURL(baseURL) + "/usage"
	req, err := http.NewRequestWithContext(ctx, "GET", endpoint, nil)
	if err != nil {
		return nil, nil
	}
	for k, v := range claudeHeaders(accessToken) {
		req.Header.Set(k, v)
	}
	resp, err := client.Do(req)
	if err != nil {
		return nil, nil
	}
	defer resp.Body.Close()
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, nil
	}
	raw, err := io.ReadAll(io.LimitReader(resp.Body, 1<<20))
	if err != nil {
		return nil, nil
	}
	var payload any
	if err := json.Unmarshal(raw, &payload); err != nil {
		return nil, nil
	}
	return ParseClaudePayload(payload, "claude", "", email, endpoint, nowMs), nil
}
